drivers/volume/portworx/portworx.go

```diff
@@ -1306,24 +1306,57 @@ func (p *portworx) pxSnapshotRestore(snapRestore *stork_crd.VolumeSnapshotRestor
 		if snapType == crdv1.PortworxSnapshotTypeCloud {
 			snapID = restoreNamePrefix + getUidforRestore(vol.Volume, restoreID)
 		}
-		var restoreErr error
-		err = wait.ExponentialBackoff(restoreAPICallBackoff, func() (bool, error) {
-			restoreErr = volDriver.Restore(vol.Volume, snapID)
-			if restoreErr != nil {
-				log.VolumeSnapshotRestoreLog(snapRestore).Warnf("In-place restore failed for %v: %v", vol.Volume, restoreErr)
-				return false, nil
-			}
-
-			return true, nil
-		})
+		// skip Restore() if volume has already under restore status
+		if vol.RestoreStatus != stork_crd.VolumeSnapshotRestoreStatusInProgress {
+			var restoreErr error
+			err = wait.ExponentialBackoff(restoreAPICallBackoff, func() (bool, error) {
+				restoreErr = volDriver.Restore(vol.Volume, snapID)
+				if restoreErr != nil {
+					log.VolumeSnapshotRestoreLog(snapRestore).Warnf("In-place restore failed for %v: %v", vol.Volume, restoreErr)
+					return false, nil
+				}
 
+				return true, nil
+			})
+			if err != nil {
+				logrus.Errorf("Unable to restore volume %v with ID %v, err %v",
+					vol.Volume, snapID, err)
+				vol.Reason = fmt.Sprintf("failed to do in-place restore, %v", restoreErr.Error())
+				vol.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusFailed
+				return err
+			}
+		}
+		log.VolumeSnapshotRestoreLog(snapRestore).Infof("checking volume state for restore status %v", vol.Volume)
+		// Get Volume Info
+		orgVol, err := volDriver.Inspect([]string{vol.Volume})
 		if err != nil {
-			logrus.Errorf("Unable to restore volume %v with ID %v, err %v",
+			return &ErrFailedToInspectVolume{
+				ID:    vol.Volume,
+				Cause: fmt.Sprintf("Volume inspect returned err: %v", err),
+			}
+		} else if len(orgVol) == 0 {
+			return &ErrFailedToInspectVolume{
+				ID:    vol.Volume,
+				Cause: fmt.Sprintf("Volume not found err: %v", err),
+			}
+		}
+
+		// check restore status
+		if orgVol[0].State == api.VolumeState_VOLUME_STATE_RESTORE {
+			// restore is in progress mark volume as stage again & return
+			if orgVol[0].Error == "" {
+				vol.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusInProgress
+				vol.Reason = "Volume restore in progress"
+				return fmt.Errorf("volume restore is in progress")
+			}
+			// restore is failed, stop and return error
+			log.VolumeSnapshotRestoreLog(snapRestore).Errorf("volume snapshot restore failed, volume %v with snap %v, err %v",
 				vol.Volume, snapID, err)
-			vol.Reason = fmt.Sprintf("failed to do in-place restore, %v", restoreErr.Error())
+			vol.Reason = fmt.Sprintf("failed to do in-place restore, %v", orgVol[0].Error)
 			vol.RestoreStatus = stork_crd.VolumeSnapshotRestoreStatusFailed
-			return err
+			return fmt.Errorf("restore failed for volume %v with err %v", vol.Volume, err)
```

Comment:
As discussed, let's wait in this function for the restore to finish instead of returning an error.

Reply:
added backoff for volume state